Add some demo attacks (CPA2 and CCA2)

ed770878 opened the pull request:
Here are some attacks to recover secret information about the key, recover secret messages, and forge encrypted messages.

This doesn't mean the algorithm is dead, it just means the algorithm is not CPA/CCA secure.  This is a stream cipher (kinda-sorta) so it's not a big surprise that message recovery and forgery are possible.  These attacks work against some advanced stream ciphers, too.

The more unsettling attacks are the ones that reveal key information.  In this case, I was able to reveal the key jumps and key length, without looking at the key.  It doesn't reveal the whole key, but it's an important first step.  The next step to key recovery might be coding the algorithm in a constraint solver, to see if it can narrow down the search space for the key body.  You would need to know the jumps and key length (for modulo) to define the constraints in the solver.

Note: these attacks all work with random data.  They don't use any property of the plain text, like it being ascii or compressed.  The attacks are equally effective for all types of data.

[Chosen Plaintext Attack (CPA)](https://en.wikipedia.org/wiki/Chosen-plaintext_attack)
[Chosen Ciphertext Attack (CCA)](https://en.wikipedia.org/wiki/Chosen-ciphertext_attack)

May I also recommend:
[Cryptography I - Stanford Online](http://online.stanford.edu/course/cryptography)

Commit:
Add some demo attacks (CPA2 and CCA2)

ikizir commented:
Sorry,

I don't know anything about Python language.
I  am not sure if I could understand your code properly.
But, I haven't seen a code which reveals any information.
Can you explain me better what that code does?

ikizir commented:
I never use nor recommend using the algorithm without Hohha Communication packet format.
Hohha Communication packet format has its own integrity check values.
Once a corrupted packet received, the server aborts the communication with the client.
It's not possible to use the server as oracle.
Are those attacks still valid in those circumstances?

ed770878 commented:
> Can you explain me better what that code does?

Yes.  I'll go write some nice long comments in the code explaining the theory of the attack, and what exactly the code does.

Python is a useful language to know.  I recommend learning the basics of it.  Lots of example code online is actually valid python code.

This attack code would look pretty much the same in c, with semicolons, and array length instead of array slice notation.  The boilerplate to run the oracle would be different (~10 times longer in c), but that's not important for understanding the attack.

> Hohha Communication packet format has its own integrity check values. Once a corrupted packet received, the server aborts the communication with the client. It's not possible to use the server as oracle. Are those attacks still valid in those circumstances?

Yes, the attacks are still possible.  For the two recovery attacks, it will take on average 128 attempts, at worst 256, to determine a valid crc for the target output.  After the target crc is known, the rest of the attack only requires one attempt per step.  If an attempt fails the hohha crc check, the error result is just as useful as a result that doesn't match the target.  When the result matches the target, the crc check will pass.  For the message forgery attack, on average 128 guesses, at worst 256, may be needed for each character.

Actually, all of the attacks would be possible, knowing **only** if the message is valid.  The attacker wouldn't need to actually observe the result of the encryption, just whether or not the server disconnects.  In the attacks against the packet format, I will just have the oracle (simulating the server) return an error if the message is invalid (simulating disconnect), no error if the message is valid, and nothing about the result of decryption.  Error or no error will be the only signal available to the attack.

I will add hohha packet format support to the harness, and write attacks against the packet format.

ikizir commented:
It's very bad that I can't undestand Python or you haven't coded in C.
I of course want to learn Python but I really have no time for that at the
moment.

Are you sure about the numbers you're giving?
In packet format, the header is also encrypted.
And CRC value is 32 bits.
I still can't understand how you can guess CRC in 128 or 256 attempts.
I even don't mention about the padding yet.

On Thu, Jan 28, 2016 at 4:06 PM, ed770878 <notifications@github.com> wrote:

> Can you explain me better what that code does?
>
> Yes. I'll go write some nice long comments in the code explaining the
> theory of the attack, and what exactly the code does.
>
> Python is a useful language to know. I recommend learning the basics of
> it. Lots of example code online is actually valid python code.
>
> This attack code would look pretty much the same in c, with semicolons,
> and array length instead of array slice notation. The boilerplate to run
> the oracle would be different (~10 times longer in c), but that's not
> important for understanding the attack.
>
> Hohha Communication packet format has its own integrity check values. Once
> a corrupted packet received, the server aborts the communication with the
> client. It's not possible to use the server as oracle. Are those attacks
> still valid in those circumstances?
>
> Yes, the attacks are still possible. For the two recovery attacks, it will
> take on average 128 attempts, at worst 256, to determine a valid crc for
> the target output. After the target crc is known, the rest of the attack
> only requires one attempt per step. If an attempt fails the hohha crc
> check, the error result is just as useful as a result that doesn't match
> the target. When the result matches the target, the crc check will pass.
> For the message forgery attack, on average 128 guesses, at worst 256, may
> be needed for each character.
>
> Actually, all of the attacks would be possible, knowing *only* if the
> message is valid. The attacker wouldn't need to actually observe the result
> of the encryption, just whether or not the server disconnects. In the
> attacks against the packet format, I will just have the oracle (simulating
> the server) return an error if the message is invalid (simulating
> disconnect), no error if the message is valid, and nothing about the result
> of decryption. Error or no error will be the only signal available to the
> attack.
>
> I will add hohha packet format support to the harness, and write attacks
> against the packet format.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ikizir/HohhaDynamicXOR/pull/13#issuecomment-176199219>
> .
>

ed770878 commented:
> In packet format, the header is also encrypted.

I think you will find this to be impossible.  In order to decrypt the message, you would need to know the salt.  If the salt is encrypted (because the header is encrypted), how will you know the what salt to use to decrypt the packet?

If the crc is encrypted as well, and the whole packet is integrity checked with crc, how do you know what crc to put in the packet, before computing the crc of the whole packet?

> I still can't understand how you can guess CRC in 128 or 256 attempts.

Because I'm not guessing the crc.  I'm guessing the character that gives that crc.  I simply compute the crc of the one character target plaintext, and then try 256 possible ciphertexts that might decode to that plaintext.  If I guess the right ciphertext character, it will decode to the target plaintext, and the crc will be valid.

> I even don't mention about the padding yet.

It is actually very simple.  Remember, the attack does not need to know anything about the message.  It is just as effective against text, compressed, random data, and even the padding.  What I would do is send a fake message with length zero, with a padding alignment of 1, and send a fake encrypted message where the pad is filled with zero (or any other byte of my choice).  I'll provide the crc value 0xad7d53ae (crc32 of the single byte zero, or other crc32 of the byte of my choice), and try 256 possible ciphertexts that might represent the fake message.  I will repeat this process for each byte of the message (always with a padding align 1, and increasing message length).  When I have recovered all of the encrypted data, I can just change the padding align to match the message.

If messages of length zero are rejected, I'll send a message of length one instead, and try 64K of possible messages to crack the first two bytes at the same time.

I can imagine your next idea might be to reject connections after N attempts.  I will distribute the attack across thousands of machines.  If the server were to reject connections from *everyone* after N attempts, it would be trivial to implement a denial of service attack against the server.  It is very hard to defend against distributed attacks, and provide service.  The service pretty much has to respond to all the requests, no matter what, because *some* of the requests *might* be from valid clients, and you still should serve those clients.  This is what makes distributed attacks so frustratingly effective.

ikizir commented:
>I think you will find this to be impossible. In order to decrypt the message, you would need to know the >salt. If the salt is encrypted (because the header is encrypted), how will you know the what salt to use >to decrypt the packet?

There are 2 salts: 
1. Key has an original salt value in the key.
2. Every encrypted packet has its own salt value in the header. The key's original salt value is solely used  for encrypting packet headers.

ikizir commented:
>It is actually very simple. Remember, the attack does not need to know anything about the message. It >is just as effective against text, compressed, random data, and even the padding. What I would do is >send a fake message with length zero, with a padding alignment of 1, and send a fake encrypted >message where the pad is filled with zero (or any other byte of my choice). I'll provide the crc value >0xad7d53ae (crc32 of the single byte zero, or other crc32 of the byte of my choice), and try 256 >possible ciphertexts that might represent the fake message. I will repeat this process for each byte of >the message (always with a padding align 1, and increasing message length). When I have recovered >all of the encrypted data, I can just change the padding align to match the message.


Padding can only be 8,16,32 or bytes.
There is no packet of 1 byte.
I am talking about this: You minimum length of the text you must guess its CRC is 16 bytes.
With your previous answer, I clearly understood that you still don't know how the algorithm works.
You make false assumptions.

ed770878 commented:
> The key's original salt value is solely used for encrypting packet headers.

This is bad.  Oh, how this is bad!  The same key AND salt are used to encrypt multiple packet headers.  Stream ciphers fail so spectacularly this way.  This offers no protection for he header.

[Stream Cipher Attack](https://en.wikipedia.org/wiki/Stream_cipher_attack)

ed770878 commented:
> you still don't know how the algorithm works

And yet, I've rewritten it, refactored it even, in a completely different programming language.  I have offered some demonstrations of how it is vulnerable, even.

Look.  The cryptographic theory always makes assumptions about what the attacker can and can not do.  What those assumptions are is stated in the attack models.  Sure, you can change the assumptions, but then you are asking for a different kind of attack.  I have shown you that *some* attacks work *assuming* the attacker can do *some things*.  If the attacker can not do those things, sure, the attack won't be effective, but if the attacker can do those things, the attack *is* effective.

ikizir commented:
>This is bad. Oh, how this is bad! The same key AND salt are used to encrypt multiple packet headers. >Stream ciphers fail so spectacularly this way. This offers no protection for he header.

I think, you are still making false assumptions.
You don't take into account the random nature of the data encrypted. CRC and random salt values every time.
And you don't take into account that the key itself dynamically updated according to plaintext.
Try to demonstrate those attacks with you, as you told me. Then, let's continue our discussion.

ikizir commented:
>Look. The cryptographic theory always makes assumptions about what the attacker can and can not >do. What those assumptions are is stated in the attack models. Sure, you can change the >assumptions, but then you are asking for a different kind of attack. I have shown you that some >attacks work assuming the attacker can do some things. If the attacker can not do those things, sure, >the attack won't be effective, but if the attacker can do those things, the attack is effective.

I know. You are right,  And I accepted when you told me the first time.
Every "base" encryption algorithm is vulnarable for some kind of attacks.
That's why, we're using padding schemes.
Without padding, most of encryption algorithms are useless, including RSA.
Even, with padding, RSA with PKCS1 is not used anymore. They recommend using OAEP.
I am also trying to learn from you. I just asked how you can attack to hohha packets.

And, if I cut the connection and force the client to change the key(That's really what I'm actually doing); you can't use the server as an oracle. You will only have 1 (or let's say 100 if you want) chances before the key expiration.

As you told, append packet attack code to harness test.
Then, let's talk again.

ikizir commented:
>And yet, I've rewritten it, refactored it even, in a completely different programming language. I have >offered some demonstrations of how it is vulnerable, even.

That's why I am surprised when you told me that's impossible to encrypt the header when I told you the header is also encrypted. These are you words:

>I think you will find this to be impossible. In order to decrypt the message, you would need to know the >salt. If the salt is encrypted (because the header is encrypted), how will you know the what salt to use >to decrypt the packet?

You refactored the code in another language and yet you don't know the header is encrypted! Let me tell you that's interesting!

Again, I will wait for your new harness code.

And, we're not here to demonstrate defense scenarios against distributed attacks. Theoretically, noone is secure against distributed denial of service attacks. Like Archimedis, everyone can move the earth with proper tool.

ed770878 commented:
> You refactored the code in another language and yet you don't know the header is encrypted! Let me tell you that's interesting!

I didn't refactor the packet code, so what's your point?  I did rewrite the cipher code, I know how it works, and that's what I've shown is vulnerable.

So, you know what?  Here is my response: **I know how your cipher algorithm works, better than you do**.  Prove me wrong.  Tell me how and why my attacks against your algorithm work.  If you know how your algorithms works better than I do, this should be no challenge for you at all.  You don't even have to look at the python code, because you know!

1. Explain the attacks against message forgery and message recovery.
    - This is the easy one.  You can just copy-paste the answer from wikipedia.
2. Explain the attack that reveals the number of key jumps.
3. Explain the attack that reveals the key length.

> I think, you are still making false assumptions.  You don't take into account the random nature of the data encrypted in header.

Again, it **does not matter that the header has random data**, because random data is just as vulnerable as anything else.  The header is encrypted with the **same key** and the **same salt** every time.  This raises a big red flag that the scheme is very likely to be vulnerable.  It's up to you to **prove** otherwise.

You need to prove that [COA](https://en.wikipedia.org/wiki/Ciphertext-only_attack), or better yet [KPA](https://en.wikipedia.org/wiki/Known-plaintext_attack) are impossible when the same key and salt are used to encrypt the header.  I already know that KPA with only one known plaintext will reveal the first byte of any other header.  I haven't ruled out that KPA might reveal the entire header, or that COA might reveal anything.

If you know your algorithm better than me, prove it.  Explain the KPA attack that reveals one byte.  This should be easy, just like (1.) above.

My proofs that the algorithm is vulnerable are easy.  I can prove vulnerability by example, by demonstrating an attack.  A demonstrated attack is impossible to refute.  Your proofs of invulnerability have to be much more rigorous.  You have to prove that an attack is **impossible**, not just that it hasn't been broken *yet*.

ikizir commented:
Thank you for your contributions.
If you have other attack samples, I will add them with pleasure.
If you or someone else proves or demonstrates with working code, a  serious attack against packet format, I will either try to develop the algorithm or cancel it.
Until that day, I will keep using this algorithm.

mastercoms commented:
@ikizir I was really interested in this project, but your unwillingness to even consider any problem or concern with your algorithm makes me wonder why you even put it on GitHub.

ed770878 commented on attacks/README.md:
```diff
@@ -0,0 +1,122 @@
+
+### Key Recovery: Key Length
+
+The key length is supposed to be secret information.  A CCA2 attack is
```
This attack is actually CCA1 (batch), not CCA2 (adaptive), because the next request to the oracle does not depend on any previous responses.  I could have generated all the requests up front, sent the batch to the oracle, and then solved for key jumps or key length without sending any more requests.  The script was written as CCA2, because it was just easier to write it that way.

ed770878 commented on attacks/README.md:
```diff
@@ -0,0 +1,122 @@
+
+### Key Recovery: Key Jumps
+
+The key number of jumps is supposed to be secret information.  A CCA2 attack is
```
This attack is actually CCA1 (batch), not CCA2(adaptive), because the next request to the oracle does not depend on any previous responses.  I could have generated all the requests up front, sent the batch to the oracle, and then solved for key jumps or key length without sending any more requests.  The script was written as CCA2, because it was just easier to write it that way.

ed770878 commented:
> Without padding, most of encryption algorithms are useless, including RSA. Even, with padding, RSA with PKCS1 is not used anymore. They recommend using OAEP.

RSA is a public key cryptosystem.  It can't really be compared to a symmetric cipher like Hohha.  One thing RSA on its own doesn't do very well: large payloads.  RSA is generally only used in combination with other algorithms like symmetric ciphers or cryptographic hash functions.  Data can be encrypted by some symmetric cipher, and then the secret key encrypted by RSA.  Data can be hashed, and the hash value signed by RSA.

PKCS1/OEAP are recommendations about how to use RSA.  They don't change the RSA algorithm.  They describe operations on plaintext, prior to encryption with RSA (or with other PK cryptosystems like RSA).  They are used, for example, to prevent an observer from seeing that two messages are the same.  Furthermore, OEAP makes it impossible to recover any of the plaintext without recovering all of it.  PKCS1/OEAP, like RSA, operate on small payloads.  They aren't applicable to stream ciphers, so they really have nothing to do with Hohha.

For stream ciphers, the Initialization Vector (IV, or salt, as we have been calling it for Hohha), is supposed to make two of the same message indistinguishable, and protect against a stream cipher attack.  IV doesn't provide an all-or-nothing guarantee like OEAP.  The all-or-nothing guarantee would not practical for a large payload.  It could be used for denial of service, because an attacker could send a very large invalid ciphertext, and the server will have no choice but to process all of it, consuming gobs of memory and cpu resources, because it won't know if the message is valid until the very end.

All of this about padding, encrypting the header, verifying the crc, etc, has to do with the packet, not the cipher.  The packet format is like what PKCS1/OEAP are for RSA, it is a "recommendation for how to use" the Hohha cipher.  The packet format is used *with* the Hohha cipher, but it is not *the* Hohha cipher.  The recommendation should be, *use it this way, otherwise the cipher is vulnerable*.  Not *the cipher is invulnerable* because of the packet format, **duh!**

The cipher *is* vulnerable to *at least* CCA/CPA.

The packet format *might mitigate* CCA/CPA (no guarantee), but it *is* vulnerable to *at least* KPA.

I am eagerly waiting to read your explanation of my attacks.

1. (attack the cipher) message forgery and message recovery
2. (attack the cipher) recover the number of jumps
3. (attack the cipher) recover the key length
4. (attack the packet format) knowing the plain text of one header, recover one byte of every other encrypted header.

ikizir commented:
Again:
And again,
And again:

At first, I thought that you are a "typical crypto wannabee" who looks at the "jargon" and apply ready remedies!
You think: 
"Oh! This is a stream cipher! What are the attack types against stream cipher?"
"Yess! They told me on Crypto 101 there is a typical attack:
https://en.wikipedia.org/wiki/Stream_cipher_attack
I'll write it to that ignorant!
He doesn't know our jargon!
And other people here, who doesn't understand about what happens, will think I'm cool!"

But I clearly understand that you're not!
And you willingly do this to mislead the people! And this is the 3rd time you're doing it:

https://en.wikipedia.org/wiki/Stream_cipher_attack
will not work on Hohha headers because:
Hohha algorithm is not a "simple xor" algorithm. It dynamically changes the key during encryption and make jumps according the input and output. Which means, unless "the same salt" and "the same plaintext" is used for encryption(which is nearly impossible); no header will produce the same encrypted header. And you can't use those values to obtain any meaningful attack information.
Read carefully(from Wikipedia):

>In other words, if anyone intercepts two messages encrypted with the same key, they can recover A >xor B, which is a form of running key cipher. Even if neither message is known, as long as both >messages are in a natural language, such a cipher can often be broken by paper-and-pencil >methods. During World War II, British cryptanalyst John Tiltman accomplished this with the Lorenz >cipher (dubbed "Tunny"). With an average personal computer, such ciphers can usually be broken in >a matter of minutes. If one message is known, the solution is trivial.

What will you obtain from the "differences" of random numbers?
And again, those numbers are not simply "xor"able; they are results of a much more complex operations with more variables involved.

You ARE TRYING TO MISLEAD THE PEOPLE!
AND THIS IS NOT THE FIRST TIME!

ikizir commented:
And about the question: Why did I put the algorithm to public domain?
I've tried to explain it on Wiki. But, I think, I must explain you again what I'm trying to do here.

As I explained in Wiki, I am not a cypto expert, nor, I want to become a crypto expert.
I created this algorithm, because, until today, nearly all standardized algorithms has been broken.
There wasn't "any" secure algorithm,  in SSL 3.0!
There were too slow in TLS 1.2.

Google thinks the same: That's why they had to choose to use ChaCha20 for encryption.
Shortly after, TLS included ChaCha20 in their drafts.

In my application, user's private chat data is encrypted twice: Optionally, it may encrypted with 4096 RSA and then encrypted with Hohha encyrption. Or, after hohha key exchange with RSA encryption, their prv. messages can be encrypted with Hohha + Hohha.  So, user has a choice.

Trying to use a new algorithm doesn't put my users privacy in danger, even if it's proven as broken. I give my users a choice.
I am enjoying this. In my projects, I like to use at least 1 "non standart" encryption. This is my choice.
I put what I could to in public domain and I left the rest to other developers imaginations.
Anyone can fork, develop, can prove what is wrong, can give new ideas.
It is a toy for developers. 

Everybody is responsible of himself or herself.
I will use Hohha as my default encryption.
If someone proves a serious vulnarability, I will try to fix it or use another encryption.

ikizir commented:
I must repeat that, you're making a very dangerous thing: You are misleading people with some false information but you also feed them with also some true information.

I must clarify the minds of readers who can't follow the conversations:

So far, I detected 2 true assumptions amongst what @ed770878  wrote:

1. The algorithm is vulnerable when used wihout Hohha Packet header. I already knew it. That's why I've developed Hohha Comm pack format. In fact, the packet format is by design of the algorithm. I designed everything from scratch for that packet format. That's why, the key contains, "one original salt" which is solely used to encrypt the packet headers. I know, this is not "on the book". But, I don't care about the jargon. I must produce working solutions and I must produce them as quick as possible. I don't care how it will be called.

It is not very serious. I don't know what the attacker can obtain but, this vulnerability is theoretically possible against packet format:

>(attack the packet format) knowing the plain text of one header, recover one byte of every other >encrypted header.

I will put one byte extra random padding byte at the first byte of the header against this attack.

Again, thank you for your contributions.
Your critics are always welcome unless you mislead the other developers.

ikizir commented:
Communcation packet format update committed.

ed770878 commented:
> And about the question: Why did I put the algorithm to public domain?

Is this public domain or MIT?  Your LICENSE is MIT.

> What will you obtain from the "differences" of random numbers?

It is a stream cipher attack.  It doesn't reveal the whole header.  It reveals the first byte of the header, but it is still *the classic* stream cipher attack.

> I will simply put one extra random padding byte as the first byte of the header against this attack which will eliminate that possibility.

And here you go again with more recommendations on how to use the algorithm securely.  I point out a vulnerability, and you say, *oh nonononono because I can put a padding byte there so it's secure, **duh!** *  That extra padding is something you can prepend to the plain text, before encrypting it with the vulnerable cipher, to mitigate that vulnerability.

> Google thinks the same: That's why they had to choose to use a non standart algorithm, ChaCha20, for encryption.

Hmm... I wonder what google thinks of hohha.  Maybe that should use that instead of ChaCha20!  I don't know where you're going with this... *RC4 wasn't good enough for google, therefore hohha is secure, QED?*

> He doesn't know our jargon! And other people here, who doesn't understand about what happens, will think I'm cool!" But I clearly understand that you're not!

What are you, a troll?  Am I being trolled?  On the internet?  It all makes sense now.  You trolled the linux mailing list to see who would bite, and I'm that sucker.

If you're serious, not trolling me, man, I feel sorry for you.  Have fun developing your algorithm, I guess.  I can't imagine how you'll ever convince anyone that they're secure, or how you'll get any help, with that attitude.  If anything is impenetrable here, it's not your algorithm, man, it's you.

I am -- still -- waiting to read your explanation of these attacks.  You really don't know how they work, do you?

1. (attack the cipher) message forgery and message recovery - of the whole message
2. (attack the cipher) recover the number of jumps
3. (attack the cipher) recover the key length
4. (attack the packet format) knowing the plain text of one header, recover one byte of every other encrypted header.
